### ISSUE TITLE
Can't login to SharePoint online on Brave

### ISSUE THREAD

**Issue by ShivanKaul** [brave/brave-browser#36450]
## Description 

Something in Brave is causing a long redirect loop when trying to login to https://unomail.sharepoint.com/. It does not look related to Shields or privacy protections, given that even with the following I still can't login:
1. Shields down for both unomail.sharepoint.com and login.microsoftonline.com
2. Tracker, fingerprinting, HTTPS upgrades, De-AMP, debouncing all set to Disabled
3. 3P cookies globally allowed

Community reports: https://community.brave.com/t/unable-to-login-into-microsoft-account-in-brave-browser-but-working-fine-for-other-browsers/534041/4

Logging into login.microsoftonline.com directly OTOH works fine.

## Steps to Reproduce

   1. Go to https://unomail.sharepoint.com/
   2. Try logging in with regular MS account

## Actual result:

![image](https://github.com/brave/brave-browser/assets/5284154/c5c6bad0-148e-4b20-8552-daf9e6eba53e)

There's a number of redirects between unomail.sharepoint.com and login.microsoftonline.com

<img width="1512" alt="image" src="https://github.com/brave/brave-browser/assets/5284154/141af42b-824a-4707-8d93-a3bd1eafc7e8">

<img width="1184" alt="image" src="https://github.com/brave/brave-browser/assets/5284154/35315aa7-749d-4126-a48c-844acc00352a">

The network activity looks like this:
1. POST to https://unomail.sharepoint.com/_forms/default.aspx returns a 302 to `https://login.microsoftonline.com:443/f1f4be86-d048-47e8-aa26-15b01dcdb13d/oauth2/authorize?client%5Fid=...`
2. GET to that redirected URL is 200
3. But then another request to https://unomail.sharepoint.com/_forms/default.aspx is started, from initiator `Other`, which gets a 302.

## Expected result:
Either logged in or a page that says that your account `can't be found in the unomail.sharepoint.com directory. Please try again later, while we try to automatically fix this for you.`

## Reproduces how often: 

Always

## Brave version (brave://version info)

1.65.6 Chromium: 122.0.6261.43 (Official Build) nightly (arm64) 
Happens on all Brave versions

**Comment by ShivanKaul** [comment]
Just to give an update here, setting Brave's User Agent to be a non-Chromium one i.e. Firefox's or Safari's seems to fix the issue.

**Comment by dovecode** [comment]
I believe I have identified the root cause of this issue. When Sharepoint initiates the authentication, it sets a cookie (from `/_forms/default.aspx` on my employer's Sharepoint -- not sure if that's standard). However, if the browser identifies as a Chromium based browser, it also clears the cookie in the same response. Essentially:

```
Set-Cookie: <cookiename>=<unique value>; expires=<timestamp>; path=/; SameSite=None; secure; HttpOnly
Set-Cookie: <cookiename>=; expires=<timestamp>; path=/; SameSite=None; Partitioned; secure; HttpOnly
```

Google Chrome reacts to this as either clearing first, then setting it, or clearing the partitioned cookie type and setting the non-partitioned one, while Brave just clears is.

As a result, post-authentication when the authentication provider redirects back to default.aspx, the cookie is missing and Sharepoint goes into a tailspin.

**Comment by ShivanKaul** [comment]
I'll take a look in a bit, but this does seem likely, we noticed some discrepancies in how cookies were being sent between Brave/Chrome/others.

**Comment by dovecode** [comment]
> I'll take a look in a bit, but this does seem likely, we noticed some discrepancies in how cookies were being sent in the request for default.aspx between Brave/Chrome/others.

Yeah. It looks like for Google Chrome if does set the cookies the same as for Brave (it probably looks at the user agent and goes "ah, Chromium, it can handle non-RFC6267-compliant cookie headers"), while for Firefox, it only sets the cookie, it doesn't clear the CHIP cookie. That's how I found I could work around the bug by spoofing Firefox.

**Comment by cb3inco** [comment]
Same issue here (Linux OS) started yesterday, but feels more like a SharePoint issue. In another profile with another Microsoft account, no issue, so it's a bit weird, unless it's because I refreshed an authentication token yesterday on the broken profile. I also tested on my Android device as well and ran into the same issue.

**Comment by ShivanKaul** [comment]
@dovecode https://github.com/brave/brave-browser/issues/36450#issuecomment-1970250241 catch looks correct to me.

For Brave (and FF), all cookies are partitioned, so a cookie being marked as Partitioned is a no-op, but because of that if you set the value of a cookie `A=1` and then in a different call `A=2; Partitioned`, from Brave's pov, it's the same cookie so the value of cookie `A` is just `1`, while in Chrome, it will have two cookies `A=1` and `A=2; Partitioned`.

I can't think of why SharePoint needs to do both clearing for the `nSGt-*` Partitioned cookie _and_ setting a value for the same unpartitioned cookie, but yeah Chrome treats that as two separate cookies while Brave just clears the cookie, causing the repeated looping. 

https://github.com/brave/brave-core/pull/22391 will fix this.

**Comment by kjozwiak** [comment]
The above requires `1.63.168` or higher for `1.63.x` verification 👍

**Comment by syaikhipin** [comment]
Same issue here with sharepoint

**Comment by mmcdougall2** [comment]
I hope Brave gets this fixed soon. I HATE M$ Edge!

**Comment by ibnuharisfawaid** [comment]
Me too, i'm having with this issue.

**Comment by goodov** [comment]
We rolled a fix for the issue via Griffin. It should be applied after a browser restart.

To check if it's there, please open `brave://version` and look for `PartitionedCookies:Enabled` line in the `Active Variations` list.

**Comment by Cyb3rN0de** [comment]
Also having this issue. Started today on my Mac only using Brave, however Brave on Kali Linux is still working fine.

**Comment by TheBule88** [comment]
> We rolled a fix for the issue via Griffin. It should be applied after a browser restart.
> 
> To check if it's there, please open `brave://version` and look for `PartitionedCookies:Enabled` line in the `Active Variations` list.

Can confirm that this fixed the issue for me!

**Comment by ibnuharisfawaid** [comment]
> We rolled a fix for the issue via Griffin. It should be applied after a browser restart.
> 
> To check if it's there, please open `brave://version` and look for `PartitionedCookies:Enabled` line in the `Active Variations` list.

Awesome, it was fixed

**Comment by cheflA1** [comment]
Same issue for me.
I don't quite understand that "fix". I would just like it to work again without all those modifications.

**Comment by AUfferte** [comment]
> Same issue for me. I don't quite understand that "fix". I would just like it to work again without all those modifications.

Restart your browser/computer
It just made the update automatically and worked for me

**Comment by LaurenWags** [comment]
Verification in progress with

```
Brave | 1.63.168 Chromium: 122.0.6261.94 (Official Build) (x86_64)
-- | --
Revision | 08196d502fa95681adc6dd9fc06a04d0664ac538
OS | macOS Version 13.6.4 (Build 22G513)
```

**Comment by hffvld** [comment]
Verified on `Pixel 7` using version(s):
```
Device/OS: Pixel 7 / panther_beta-user 14 AP21.240119.009 release-keys
Brave build: 1.63.168
Chromium: 122.0.6261.94 (Official Build) (64-bit) 
```
---------------------------

**STEPS:**
1. Launch Brave
2. Go to https://unomail.sharepoint.com/
3. Try logging in with a regular MS account > *Verify*

**ACTUAL RESULTS:**
- Verified that the page says `That didn't work` and your account `can't be found in the unomail.sharepoint.com directory` are shown when signing in with a regular MS account to https://unomail.sharepoint.com/

---------------------------

|1|2|
|:-:|:-:|
| ![1](https://github.com/brave/brave-browser/assets/128532965/2cba6bfe-4535-41e4-a4fc-f11d507b5fd1) | ![2](https://github.com/brave/brave-browser/assets/128532965/c31327c3-6b16-4d5f-a7d3-63757cb7b58e) |

**Comment by Uni-verse** [comment]
Verified on `Samsung Galaxy Tab S7` using version:

```
Brave	1.63.168 Chromium: 122.0.6261.94 (Official Build) (64-bit) 
Revision	08196d502fa95681adc6dd9fc06a04d0664ac538
OS	Android 13; Build/TP1A.220624.014; 33; REL
```

- Ensured logging into https://unomail.sharepoint.com/ will present `That didn't work` page instead of `something went wrong`
- Ensured ephemeral storage tests in https://dev-pages.brave.software/storage/ephemeral-storage.html for Brave pre-1.64.31 are successfully passed. 

    - [x] Cross-site cookies blocked
    - [x] Cookies blocked
    - [x] All cookies allowed
    
1.63.165 | 1.63.68 | version
-|-|-  
![Screenshot 2024-03-05 at 12 53 49 PM](https://github.com/brave/brave-browser/assets/17885425/02ef3042-7622-48c3-a016-b7e247a5ec53) | ![Screenshot 2024-03-05 at 12 55 54 PM](https://github.com/brave/brave-browser/assets/17885425/1112490d-0bbc-4217-9612-0e042aa43ca6) | ![Screenshot 2024-03-05 at 12 57 19 PM](https://github.com/brave/brave-browser/assets/17885425/e49dac97-4fac-49c6-adf8-265f944078a6)

**Comment by MadhaviSeelam** [comment]
Verification `PASSED` using
```
Brave | 1.63.165 Chromium: 122.0.6261.94 (Official Build) (64-bit)
-- | --
Revision | fed0fcc915f6d9e1db871202f0304c8d582a73de
OS | Windows 11 Version 23H2 (Build 22631.3155)
```

Reproduced the issue in `1.63.162 Chromium: 122.0.6261.69`
![image](https://github.com/brave/brave-browser/assets/98358127/e6000667-949e-4de3-85c6-fc04353478a4)

1. Installed 1.63.165
2. launched Brave
3. logged into https://unomail.sharepoint.com/ with Microsoft account
4. confirmed `That didn't work... can't be found in the unomail.sharepoint.com directory` message shown as expected since I do not have  Sharepoint account.

![Screenshot 2024-03-05 123457](https://github.com/brave/brave-browser/assets/98358127/cf08d16b-8c90-4225-999a-df71c3b07e3c)

 https://dev-pages.brave.software/storage/ephemeral-storage.html 
verified the test cases from https://dev-pages.brave.software/storage/ephemeral-storage.html for Brave pre-1.64.31+ with Chromium storage partitioning.

* `Cross-site cookies blocked` - **PASSED**
* `Cookies Blocked` - **PASSED**
* `All cookies allowed` - **PASSED**

**Comment by kjozwiak** [comment]
> cc @kjozwiak for the remaining check on macOS

Verification **PASSED** on `macOS Senoma 14.3.1 x64` using the following build(s): 

```
Brave | 1.63.168 Chromium: 122.0.6261.94 (Official Build) (x86_64)
--- | ---
Revision | 08196d502fa95681adc6dd9fc06a04d0664ac538
OS | macOS Version 14.3.1 (Build 23D60)
```

Visited `https://unomail.sharepoint.com` and ensured that the error displays `That didn't work` and not `Something went wrong` as mentioned via https://github.com/brave/brave-browser/issues/36450#issue-2160056430.

![Screenshot 2024-03-05 at 4 23 30 PM](https://github.com/brave/brave-browser/assets/2602313/76bed25c-84e7-4c1a-a296-e0368657c61a)

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Ubuntu 22.04 x64` using the following build(s):

```
Brave | 1.63.168 Chromium: 122.0.6261.94 (Official Build) (64-bit)
--- | ---
Revision | 08196d502fa95681adc6dd9fc06a04d0664ac538
OS | Linux
```

Using the STR/information detailed via https://github.com/brave/brave-browser/issues/36450#issue-2160056430, ensured that logging into `https://unomail.sharepoint.com` produced an error message as I don't have a Sharepoint account rather than failing via `Something went wrong` as per the following:

![image](https://github.com/brave/brave-browser/assets/2602313/f7bea734-5309-4517-ae42-8d41b40940fe)

Also went through https://dev-pages.brave.software/storage/ephemeral-storage.html and verified that all the tests passed as per:

Using `Brave 1.64.31+ with Chromium storage partitioning`, ensured all three tests/shield states passed as per:

* `Cross-site cookies blocked` - **PASSED**
* `Cookies Blocked` - **PASSED**
* `All cookies allowed` - **PASSED**

**Comment by mt-drunkdragon** [comment]
It seems to be fixed without Bave update.  Mine is 1.63.169.
Maybe MS did something...

**Comment by mephinet** [comment]
Can confirm: upgrading to 1.63.169 fixes the issue. Thanks for the quick solution!

**Comment by dovecode** [comment]
> It seems to be fixed without Bave update. Mine is 1.63.169. Maybe MS did something...

Nah, 1.63.169 *is* the upgrade that works around MS's non-RFC6267-compliant cookies. I can confirm that Sharepoint still sends two Set-Cookie headers with the same cookie name.

**Comment by mmcdougall2** [comment]
Mine works now, but I don't know what changed.
Thanks!

***@***.***<http://www.dallascitynews.net/>



Mike McDougall
 Sr. Systems Admin
 City of Dallas | DallasCityNews.net
  CIS
  1500 Marilla, 3ES
  Dallas, TX 75201
  O:  214-670-3466
 ***@***.******@***.***>
   ***@***.*** <https://twitter.com/CityofDallas>  ***@***.*** <https://www.facebook.com/DallasCityHall/>   ***@***.*** <https://www.youtube.com/user/dmcclel>

**OPEN RECORDS NOTICE: This email and responses may be subject to the Texas Open Records Act and may be disclosed to the public upon request.  Please respond accordingly.**

**Comment by jd2476** [comment]
I still randomly have this issue.  The workaround I found was clearing cookies and other site data, which is not practical as I need to log into all my accounts again.

The error I get when trying to load a company Sharepoint page is:

The webpage at <url> might be temporarily down or it may have moved permanently to a new web address.

Running macOS Sonoma 14.4.1 (this issue was prevalent ever since I upgraded to Ventura) and Brave Version 1.65.126 Chromium: 124.0.6367.118 (Official Build) (arm64)
